qiskit_experiments/analysis/curve_analysis.py

```diff
@@ -0,0 +1,764 @@
+                    "match with the length of expected parameters. "
+                    f"{len(parameter_array)} != {len(fit_params)}."
+                )
+            return dict(zip(fit_params, parameter_array))
+
+        if "p0" in fitter_options:
+            if isinstance(fitter_options["p0"], dict):
+                _check_keys("p0")
+            else:
+                fitter_options["p0"] = _dictionarize("p0")
+        else:
+            raise AnalysisError("Initial guess p0 is not provided to the fitting options.")
+
+        if "bounds" in fitter_options:
+            if isinstance(fitter_options["bounds"], dict):
+                _check_keys("bounds")
+            else:
+                fitter_options["bounds"] = _dictionarize("bounds")
+        else:
+            fitter_options["bounds"] = dict(zip(fit_params, [(-np.inf, np.inf)] * len(fit_params)))
+
+        return fitter_options
+
+    def _subset_data(
+        self,
+        name: str,
+        series: np.ndarray,
+        x_values: np.ndarray,
+        y_values: np.ndarray,
+        y_sigmas: np.ndarray,
+    ) -> Tuple[np.ndarray, ...]:
+        """A helper method to extract reduced set of data.
+
+        Args:
+            name: Series name to search for.
+            series: An integer array representing a mapping of data location to series index.
+            x_values: Full data set of x values.
+            y_values: Full data set of y values.
+            y_sigmas: Full data set of y sigmas.
+
+        Returns:
+            Tuple of x values, y values, y sigmas for the specific series.
+
+        Raises:
+            AnalysisError:
+                - When name is not defined in the __series__ definition.
+        """
+        for idx, series_def in enumerate(self.__series__):
+            if series_def.name == name:
+                data_index = series == idx
+                return x_values[data_index], y_values[data_index], y_sigmas[data_index]
+        raise AnalysisError(f"Specified series {name} is not defined in this analysis.")
+
+    def _run_analysis(
+        self, experiment_data: ExperimentData, **options
+    ) -> Tuple[List[AnalysisResult], List["pyplot.Figure"]]:
+        """Run analysis on circuit data.
+
+        Args:
+            experiment_data: the experiment data to analyze.
+            options: kwarg options for analysis function.
+
+        Returns:
+            tuple: A pair ``(analysis_results, figures)`` where
+                   ``analysis_results`` may be a single or list of
+                   AnalysisResult objects, and ``figures`` is a list of any
+                   figures for the experiment.
+
+        Raises:
+            AnalysisError: if the analysis fails.
+        """
+        analysis_result = CurveAnalysisResult()
+        figures = list()
+
+        # pop arguments that are not given to fitter
+        curve_fitter = options.pop("curve_fitter")
+        data_processor = options.pop("data_processor")
```

Comment:
This means that we need to add a data processor to the default run options. This is fine. But if the user overrides the default options (e.g. requesting `KERNELED` data instead of `CLASSIFIED` data) then he also needs to adjust the data processor accordingly which may lead to inconsistencies. Is this desired?

Reply:
Yes, but I think this is not the problem only spectroscopy or calibration experiments run into. In 0028669 I added temporarily logic to dynamically update this option, but I believe we should update the processor logic. Basically circuit result contains these measurement options in the header, so if the all processors support data processing from the level1 data and flexibly change starting node wrt the measurement conditions attached to `data`, we don't need to worry about this problem.

Reply:
For example, user may want to run QPT with level1 measurement to see the effect of leakage. If the probability processor can support both level1 and 2, we don't need to switch processor depending on input.